File: web-processor/src/main/java/cn/labzen/web/apt/evaluate/generics/SimplestControllerInterfaceGenericsEvaluator.java
package cn.labzen.web.apt.evaluate.generics;

import cn.labzen.web.apt.internal.context.AnnotationProcessorContext;
import cn.labzen.web.apt.suggestion.Suggestion;
import com.squareup.javapoet.TypeName;

import java.util.List;

import static cn.labzen.web.apt.definition.TypeNames.INTERFACE_SIMPLEST_CONTROLLER;

/**
 * 对应  cn.labzen.web.controller.SimplestController
 */
public final class SimplestControllerInterfaceGenericsEvaluator extends PrimaryServiceGenericsAssign {

  @Override
  protected String supportedInterfaceName() {
    return INTERFACE_SIMPLEST_CONTROLLER;
  }

  @Override
  public List<? extends Suggestion> evaluate(List<TypeName> arguments) {
    assert !arguments.isEmpty();
    return internalEvaluate(arguments.getFirst());
  }
}

File: src/main/kotlin/cn/labzen/web/ap/evaluate/generics/SimplestControllerInterfaceGenericsEvaluator.kt
package cn.labzen.web.ap.evaluate.generics

import cn.labzen.web.ap.suggestion.Suggestion
import cn.labzen.web.controller.SimplestController
import com.squareup.javapoet.ParameterizedTypeName
import com.squareup.javapoet.TypeName

/**
 * 对应 [SimplestController]
 */
class SimplestControllerInterfaceGenericsEvaluator : PrimaryServiceGenericAssign(), InterfaceGenericsEvaluator {

  override fun support(type: TypeName): Boolean =
    if (type is ParameterizedTypeName) {
      TYPE == type.rawType
    } else {
      TYPE == type
    }

  override fun evaluate(arguments: List<TypeName>): List<Suggestion> {
    return internalEvaluate(arguments.first())
  }

  companion object {
    private val TYPE = TypeName.get(SimplestController::class.java)
  }
}